src/cli/test/reporters/github/index.ts

```diff
@@ -0,0 +1,66 @@
+import { Context, Reporter } from '@jest/reporters';
+import { AggregatedResult, TestResult } from '@jest/test-result';
+import pMap from 'p-map';
+
+import * as GitHub from '../../../../api/github';
+import {
+  buildNameFromEnvironment,
+  enabledFromEnvironment,
+} from '../../../../api/github/environment';
+
+import { createAnnotations } from './annotations';
+
+const DEFAULT_NAME = 'noDisplayName';
+
+export default class GitHubReporter implements Pick<Reporter, 'onRunComplete'> {
+  async onRunComplete(
+    _contexts: Set<Context>,
+    results: AggregatedResult,
+  ): Promise<void> {
+    if (!enabledFromEnvironment()) {
+      return;
+    }
+
+    // Sort tests by display name
+    const sortedTestResults: Record<string, TestResult[]> = {};
+    results.testResults.forEach((testResult) => {
+      const displayName = testResult.displayName?.name ?? DEFAULT_NAME;
+      sortedTestResults[displayName] ??= [];
+      sortedTestResults[displayName].push(testResult);
+    });
+
+    // Create annotations for each display name
+    const annotationResults = Object.entries(sortedTestResults).map<{
+      displayName: string;
+      annotations: GitHub.Annotation[];
+    }>(([displayName, testResults]) => ({
+      displayName,
+      annotations: createAnnotations(testResults),
+    }));
+
+    // Create a check run per display name. Run in series.
+    await pMap(
+      annotationResults,
+      async ({ displayName, annotations }) => {
+        const name = `skuba/test${
+          displayName !== DEFAULT_NAME ? ` (${displayName})` : ''
+        }`;
+        const isOk = Boolean(!annotations.length);
+        const conclusion = isOk ? 'success' : 'failure';
+        const summary = isOk
+          ? '`skuba test` passed.'
+          : '`skuba test` found issues that require triage.';
+        const build = buildNameFromEnvironment();
+
+        await GitHub.createCheckRun({
+          name,
+          annotations,
+          conclusion,
+          summary,
+          title: `${build} ${isOk ? 'passed' : 'failed'}`,
```

Comment:
Not for this PR, but I'd be interested to see whether we can extend this with coverage in future. It would be neat if we could compare against the default branch a la [Codecov](https://docs.codecov.com/docs/pull-request-comments); could we try to retrieve details of the equivalent check run on the head commit for comparison?

Reply:
Might be able to grab it from the test result https://github.com/facebook/jest/blob/8f2cdad7694f4c217ac779d3f4e3a150b5a3d74d/packages/jest-test-result/src/types.ts#L95